rllib/agents/alpha_star/alpha_star.py

```diff
@@ -0,0 +1,592 @@
+                actors=self.workers.remote_workers()
+                or [self.workers.local_worker()],
+                ray_wait_timeout_s=0.01,
+                max_remote_requests_in_flight_per_actor=2,
+                remote_fn=self._sample_and_send_to_buffer,
+            )
+        # Update sample counters.
+        for (env_steps, agent_steps) in sample_results.values():
+            self._counters[NUM_ENV_STEPS_SAMPLED] += env_steps
+            self._counters[NUM_AGENT_STEPS_SAMPLED] += agent_steps
+
+        # Trigger asynchronous training update requests on all learning
+        # policies.
+        with self._timers[LEARN_ON_BATCH_TIMER]:
+            pol_actors = []
+            args = []
+            for i, (pid, pol_actor,
+                    repl_actor) in enumerate(self.distributed_learners):
+                pol_actors.append(pol_actor)
+                args.append([repl_actor, pid])
+            train_results = asynchronous_parallel_requests(
+                remote_requests_in_flight=self.remote_requests_in_flight,
+                actors=pol_actors,
+                ray_wait_timeout_s=0.1,
+                max_remote_requests_in_flight_per_actor=2,
+                remote_fn=self._update_policy,
+                remote_args=args,
+            )
+
+        # Update sample counters.
+        for result in train_results.values():
+            if NUM_AGENT_STEPS_TRAINED in result:
+                self._counters[NUM_AGENT_STEPS_TRAINED] += result[
+                    NUM_AGENT_STEPS_TRAINED]
+
+        # For those policies that have been updated in this iteration
+        # (not all policies may have undergone an updated as we are
+        # requesting updates asynchronously):
+        # - Gather train infos.
+        # - Update weights to those remote rollout workers that contain
+        #   the respective policy.
+        with self._timers[SYNCH_WORKER_WEIGHTS_TIMER]:
+            train_infos = {}
+            policy_weights = {}
+            for pol_actor, policy_result in train_results.items():
+                if policy_result:
+                    pid = self.distributed_learners.get_policy_id(pol_actor)
+                    train_infos[pid] = policy_result
+                    policy_weights[pid] = pol_actor.get_weights.remote()
+
+            policy_weights_ref = ray.put(policy_weights)
+
+            for worker in self.workers.remote_workers():
+                worker.set_weights.remote(policy_weights_ref)
+
+        return train_infos
+
+    def build_league(self, result: ResultDict) -> None:
+        """Method containing league-building logic. Called after train step.
+
+        Args:
+            result: The most recent result dict with all necessary stats in
+                it (e.g. episode rewards) to perform league building
+                operations.
+        """
+
+        # If no evaluation results -> Use hist data gathered for training.
```

Comment:
I really think we need to define some kind of API for this league building thing.
it could be as simple as
```
class League:
    def build(agent: Trainer, result; ResultDict) -> PolicyMapping:
        ...
```
we then separate all these logic in a different file, and plug in this League building object when we construct AlphaStar agent.
this would allow people to easily override the league building strategy without touch this agent.

Reply:
Yeah, I like this idea, too. Will separate.

Reply:
Done, created a LeagueBuilder base class, from which a AlphaStarLeagueBuilder sub-classes. Class and c'tor kwarg values can be specified in the config. Also works in yaml files as done in the 4-agent CartPole example.